lib/src/rules/prefer_final_fields.dart

```diff
@@ -135,21 +152,44 @@ class _Visitor extends SimpleAstVisitor<void> {
       return;
     }
 
-    fields.variables.forEach((VariableDeclaration variable) {
+    for (var variable in fields.variables) {
       final element = variable.declaredElement;
-      if (!element.isPrivate) {
-        return;
-      }
-
-      if (variable.initializer == null) {
-        return;
-      }
 
-      if (_mutatedFields.contains(element)) {
-        return;
+      if (element.isPrivate && !_mutatedFields.contains(element)) {
+        bool fieldInConstructor(constructor) =>
+            constructor.initializers.any((initializer) =>
+                _containedInInitializer(element, initializer)) ||
+            constructor.parameters.parameters
+                .any((formal) => _containedInFormal(element, formal));
+
+        final classDeclaration = node.parent;
+        final constructors = (classDeclaration is ClassDeclaration
+            ? classDeclaration.members
+                .whereType<ConstructorDeclaration>()
+                .toSet()
+            : <ConstructorDeclaration>{});
+        final isFieldInConstructors = constructors.any(fieldInConstructor);
+        final isFieldInAllConstructors = constructors.every(fieldInConstructor);
+
+        if (isFieldInConstructors) {
+          if (isFieldInAllConstructors) {
+            rule.reportLint(variable);
+          }
+        } else if (element.initializer != null) {
+          rule.reportLint(variable);
+        }
       }
-
-      rule.reportLint(variable);
-    });
+    }
   }
 }
+
+bool _containedInInitializer(
+        Element element, ConstructorInitializer initializer) =>
+    initializer is ConstructorFieldInitializer &&
+    DartTypeUtilities.getCanonicalElementFromIdentifier(
+            initializer.fieldName) ==
+        element;
+
+bool _containedInFormal(Element element, FormalParameter formal) =>
+    formal is FieldFormalParameter &&
+    formal.declaredElement.toString() == element.toString();
```

Comment:
I don't think it's necessary to convert the elements to strings before comparing them, and might open the door for a false positive if two elements that are not equal happen to generate the same string.

Reply:
I was actually out of ideas on how to get to the FieldElement from the FormalParameter but I digging a bit deeper I found out how.